include/networkit/graph/Graph.hpp

```diff
@@ -1363,7 +1363,151 @@ class Graph final {
      */
     Graph(count n = 0, bool weighted = false, bool directed = false, bool edgesIndexed = false);
 
-    Graph(const Graph &G, bool weighted, bool directed, bool edgesIndexed = false);
+    template <class EdgeMerger = std::plus<edgeweight>>
+    Graph(const Graph &G, bool weighted, bool directed, bool edgesIndexed = false,
+          EdgeMerger edgeMerger = std::plus<edgeweight>())
+        : n(G.n), m(G.m), storedNumberOfSelfLoops(G.storedNumberOfSelfLoops), z(G.z),
+          omega(edgesIndexed ? G.omega : 0), t(G.t), weighted(weighted), directed(directed),
+          edgesIndexed(edgesIndexed), // edges are not indexed by default
+          exists(G.exists),
+
+          // let the following be empty for the start, we fill them later
+          inEdges(0), outEdges(0), inEdgeWeights(0), outEdgeWeights(0), inEdgeIds(0), outEdgeIds(0),
+
+          // empty node attribute map as last member for this graph
+          nodeAttributeMap(this) {
+
+        if (G.isDirected() == directed) {
+            // G.inEdges might be empty (if G is undirected), but
+            // that's fine
+            inEdges = G.inEdges;
+            outEdges = G.outEdges;
+
+            // copy weights if needed
+            if (weighted) {
+                if (G.isWeighted()) {
+                    // just copy from G, again either both graphs are directed or both are
+                    // undirected
+                    inEdgeWeights = G.inEdgeWeights;
+                    outEdgeWeights = G.outEdgeWeights;
+                } else {
+                    // G has no weights, set defaultEdgeWeight for all edges
+                    if (directed) {
+                        inEdgeWeights.resize(z);
+                        for (node u = 0; u < z; u++) {
+                            inEdgeWeights[u].resize(G.inEdges[u].size(), defaultEdgeWeight);
+                        }
+                    }
+
+                    outEdgeWeights.resize(z);
+                    for (node u = 0; u < z; u++) {
+                        outEdgeWeights[u].resize(outEdges[u].size(), defaultEdgeWeight);
+                    }
+                }
+            }
+            if (G.hasEdgeIds() && edgesIndexed) {
+                inEdgeIds = G.inEdgeIds;
+                outEdgeIds = G.outEdgeIds;
+            }
+        } else if (G.isDirected()) {
+            // G is directed, but we want an undirected graph
+            // so we need to combine the out and in stuff for every node
+            outEdges.resize(z);
+            if (weighted)
+                outEdgeWeights.resize(z);
+            if (G.hasEdgeIds() && edgesIndexed)
+                outEdgeIds.resize(z);
+            G.balancedParallelForNodes([&](node u) {
```

Comment:
I'd recommend using `parallelForNodes`. The workload depends on the degree of `u` and with the `guided` omp schedule we would incur the same inefficiencies as described in #819.

Reply:
I benchmarked both the `parallelForNodes` and `balancedParallelForNodes` and `balancedParallelForNodes` ended being 20-40% faster.

Reply:
Could you provide further details about those benchmarks? As I did in #819, It would be helpful to know the size of the graphs, their degree distribution, the running time of the two loops, and the machine you used.

Reply:
Thanks. Compared to #819 these are rather small graphs and the balanced version saves us less than 2s in the best case. Can you run the same experiment on some larger graphs (e.g., up to a few billion edges) on a more powerful server and see if the results change?

Reply:
I tried with some bigger graphs on the gruenau1 server and the trend of `balancedParallelForLoops` being faster continues. I will try with graphs bigger than 1 billion edges next to see if it changes something.

| graph name | Nodes | Edges | Time with parallelForLoops in us| Time with balancedParallelForLoops |
| :---         |     :---:      |      :---:      |          ---: |          ---: |
| [web-indochina-2004-all](https://networkrepository.com/web-indochina-2004-all.php)   | 7.5M     | 194M   |1 342 921 834|66 780 383|
| [web-uk-2002-all](https://networkrepository.com/web-uk-2002-all.php)     | 18.5M    | 298M   |40 839 508| 29 361 124|

Reply:
Thanks for running the experiments. I forgot to mention that in #819 the recommendation is to use the `dynamic` schedule. `parallelForNodes` uses the default static schedule which, as your experiments are showing, seems to be worse than the `guided` one for this use case. I think that for this PR it is fine to use `balancedParallelForNodes`.